tachyon/crypto/sumcheck/sumcheck_verifier.h

```diff
@@ -0,0 +1,267 @@
+// Copyright 2022 arkworks contributors
+// Use of this source code is governed by a MIT/Apache-2.0 style license that
+// can be found in the LICENSE-MIT.arkworks and the LICENCE-APACHE.arkworks
+// file.
+
+#ifndef TACHYON_CRYPTO_SUMCHECK_SUMCHECK_VERIFIER_H_
+#define TACHYON_CRYPTO_SUMCHECK_SUMCHECK_VERIFIER_H_
+
+#include <utility>
+#include <vector>
+
+#include "tachyon/crypto/sumcheck/sumcheck_prover_msg.h"
+#include "tachyon/crypto/sumcheck/sumcheck_verifier_msg.h"
+#include "tachyon/crypto/sumcheck/sumcheck_verifying_key.h"
+
+namespace tachyon::crypto {
+
+template <typename F>
+F InterpolateUniPoly(const std::vector<F>& poly, const F& evaluation_point);
+
+// Subclaim created when verifier is convinced.
+template <typename F>
+struct Subclaim {
+  // The multi-dimensional point this multilinear evaluation is evaluated at.
+  std::vector<F> point;
+  // The expected evaluation.
+  F expected_evaluation;
+};
+
+template <typename MLE>
+class SumcheckVerifier {
+ public:
+  using F = typename MLE::Field;
+
+  SumcheckVerifier() = default;
+  explicit SumcheckVerifier(const SumcheckVerifyingKey& key)
+      : num_variables_(key.num_variables),
+        max_evaluations_(key.max_evaluations) {
+    randomness_.reserve(key.num_variables);
+    polynomials_received_.reserve(key.num_variables);
+  }
+  explicit SumcheckVerifier(SumcheckVerifyingKey&& key)
+      : num_variables_(key.num_variables),
+        max_evaluations_(key.max_evaluations) {
+    randomness_.reserve(key.num_variables);
+    polynomials_received_.reserve(key.num_variables);
+  }
+
+  // Runs a "round" given a |SumcheckProverMsg|, sampling and storing a random
+  // value.
+  //
+  // While a true verifier round should perform actual verification, |Round()|
+  // merely samples and stores random values. Verifications will be performed
+  // altogether in |CheckAndGenerateSubclaim| after all prover and verifier
+  // sub-rounds are subsequently finished.
+  //
+  // Referencing https://people.cs.georgetown.edu/jthaler/sumcheck.pdf
+  // |random_value| is the challenge that will given from the Verifier
+  // to the Prover aka rᵢ.
+  template <size_t MaxDegree>
+  SumcheckVerifierMsg<F> Round(SumcheckProverMsg<F, MaxDegree>&& prover_msg) {
+    CHECK_LE(round_, num_variables_);
+
+    const F random_value = F::Random();
+    randomness_.push_back(random_value);
+    polynomials_received_.push_back(
+        std::move(prover_msg.evaluations.evaluations()));
+
+    ++round_;
```

Comment:
```suggestion
    CHECK_LE(round_++, num_variables_);

    const F random_value = F::Random();
    randomness_.push_back(random_value);
    polynomials_received_.push_back(
        std::move(prover_msg.evaluations.evaluations()));
```
How about like this? cc.@chokobole

Reply:
I like it! but IMO the check should be modified like `CHECK_LE(++round_, num_variables_);`